### ISSUE TITLE
Inconsistent GPU model results

### ISSUE THREAD

**Issue by rgstephens** [RasaHQ/rasa#9737]
### Rasa Open Source version

2.8.2

### Rasa SDK version

 _No response_ 

### Rasa X version

 _No response_ 

### Python version

3.7

### What operating system are you using?

Linux

### What happened?

Results generating a model from the same training data can be significantly different when built on a GPU from one build to the next.

Details can be seen [here](https://github.com/RasaHQ/customers/issues/209).

Slack conversations regarding this issue can be viewed [here](https://rasa-hq.slack.com/archives/C021ZTVSY77/p1631729965379300) and [here](https://rasa-hq.slack.com/archives/C021ZTVSY77/p1632795613008000).

### Command / Request

 _No response_ 

### Relevant log output

 _No response_

**Comment by tttthomasssss** [comment]
@rgstephens can you provide some more context for reproducing this issue? I will start an investigation but to speed things up I'd need:
* Whats the best branch from the NIB repo to reproduce this (I'll start from `master`)
* Which data is used to train the model (there are 3 folders inside `data` - `common`, `nibau-voice`, `nihau-web`)
* There is a training script in the repo, executing the following command - is this the one that should reproduce the observed behaviour:
    ```bash
        rasa train \
          --data /opt/ml/input/data/train \
          --config /opt/ml/config.yml \
          --domain /opt/ml/data/domain \
          --out /opt/ml/model \
          --augmentation 49 \
          --fixed-model-name ${current_time}-${BOT} \
          ${FORCE_TRAIN:+ --force}
    ```
* Whats the main problem that should be investigated? The issue in [NIB#209](https://github.com/RasaHQ/customers/issues/209) sounds as if fluctuations in model confidence values is the main problem (because its hard to come up with a good fallback threshold). So is it fluctuations in model confidences or fluctuations in model performance (i.e. intent classification F1/Acc), or both?
* Has this issue been observed with a previous Rasa version (i.e. `<2.8`) and/or a previous TF version (i.e. `<2.3`)?

**Comment by tttthomasssss** [comment]
@rgstephens there are still a number of unaddressed sources of non-determinism with TF GPU training (see [here](https://github.com/NVIDIA/framework-determinism/blob/master/tensorflow_status.md)) .

While setting `TF_DETERMINISTIC_OPS: 1` didn't do the job - can you check whether the following has been tried as well (also from [this page](https://github.com/NVIDIA/framework-determinism/blob/master/tensorflow_status.md), just under the Summary table):

* **TF_DETERMINISTIC_OPS**: Set environment variable `TF_DETERMINISTIC_OPS` to `'1'` or `'true'`. Also _do not_ set environment variable `TF_USE_CUDNN_AUTOTUNE` at all (and particularly _do not_ set it to `'0'` or `'false'`).
* **TF_CUDNN_DETERMINISTIC**: Set environment variable `TF_CUDNN_DETERMINISTIC` to `'1'` or `'true'`. Also _do not_ set environment variable `TF_USE_CUDNN_AUTOTUNE` at all (and particularly _do not_ set it to `'0'` or `'false'`).

**Comment by akelad** [comment]
Rasa OSS 2.8 doesn't have 2.4 as a requirement though, as far as I know you can't just upgrade to tf 2.4 just like that - https://github.com/RasaHQ/rasa/issues/7619#issue-772336870

**Comment by koernerfelicia** [comment]
> for unsupported types such as tf.float64 an exception will be raised if tf.sparse.sparse_dense_matmul is used with the env var TD_DETERMINISTIC_OPS switched on (see the TF PR here) and the TF 2.6 changelog.

@akelad @rgstephens we've actually run into this exception being thrown when `TF_DETERMINISTIC_OPS` is set with TF 2.6 **even though when we use tf.float32**. It turns out that the deterministic implementation (for any data dtype, including float32) is **not** available for standard TF. If users/customers want determinism, they will have to use the NVIDIA container `NGC TF 21.04` onwards.

As Thomas noted above, we never had determinism for this particular op. As far as I understand, it was briefly possible for TF 2.4-2.5, but no longer in 2.6. Is this something we want to support? If so, we'd need to look into the NVIDIA container further. I'd create a followup issue and try to get the relevant eyes on it.

**Comment by rgstephens** [comment]
@JEM-Mosig Can you explain further about the `validation set` that would be needed? This [private repo](https://github.com/rasahq/nib) is a well tested production bot which runs cross validation in their ci/cd pipeline as we recommend.

**Comment by JEM-Mosig** [comment]
@rgstephens Cross-validation should also do it. But if a model generates wildly different outputs when trained with different random seeds, then either the training data isn't good or it hasn't converged (or both). But that should show up in cross-validation. Or is this mostly about confidence values?

**Comment by akelad** [comment]
@TyDunn what's happening with this issue?

**Comment by TyDunn** [comment]
@akelad Followed up in Slack [here](https://rasa-hq.slack.com/archives/C021ZTVSY77/p1635779598085900?thread_ts=1634577418.257700&cid=C021ZTVSY77)

**Comment by rgstephens** [comment]
Any updates on this issue?  I may be running into this with another bot.

**Comment by dakshvar22** [comment]
Update:
Fluctuations in performance of models in terms of F1 score on test set was investigated in this [issue ](https://github.com/RasaHQ/research/issues/269). Results were documented [internally](https://www.notion.so/rasa/Observed-fluctuations-in-model-performance-when-trained-on-GPU-dedda6d012bc489cb358e9b010903765) and in the [official docs](https://rasa.com/docs/rasa/next/tuning-your-model#deterministic-operations) as well. (Sneak peek: Extremely small fluctuations observed, the customer dataset was also included in the investigation.)

What hasn't been investigated:

As this [issue](https://github.com/RasaHQ/customers/issues/209) mentions, the problem might be one level deeper in the values of confidences that are predicted by the models. Since the former issue did not investigate this, we should look into that as part of this issue.

One hunch that I have: The config used in the dataset has `ConveRTFeaturizer` included. The featurizer uses batch inference which might result in [this bug](https://rasa-hq.slack.com/archives/C8LEHMTAM/p1639139934223600) since `tensorflow-text` might be using `tf.einsum` operation. So, the real problem might be with the featurizer.

@rgstephens If you have time, can you re-run your analysis with two different configs:

1. `ConveRTFeaturizer` removed, hence keeping only the sparse featurizers.
2. Use `LanguageModelFeaturizer` as the featurizer in pipeline (no other dense / sparse featurizer). Recommending this because using only dense features will reduce the fluctuation drastically (the investigation showed us that) and `LanguageModelFeaturizer` can get comparable results to `ConveRTFeaturizer`. The increased memory need is an issue but we might be able to confirm the hunch I mentioned above.
